### ISSUE TITLE
Stream generated output to PSR7 response

### ISSUE THREAD

**Issue by pavarnos** [PHPOffice/PHPWord#1578]
This is:

- [ ] a feature request

* PHP version: 7.2
* PHPWord version: 0.16

Is there a way to get the PhpWord Writer to return a stream that can be used to build a PSR7 Response? Otherwise we end up with two temporary copies of the file contents eg 
- the Word2007 writer creates a temp file for the zip archive and then cleans it up inside the class 
- then to stream the file to the user, i need a separate temp file to hold the content because `php://output` can not be used
and all this holds up the server because I need to delete the temp file after the request is done.
eg i need code like
```php
        $tempFile = tempnam(VAR_PATH, $fileName);
        $writer = \PhpOffice\PhpWord\IOFactory::createWriter($file, 'Word2007');
        $writer->save($tempFile);
        $response = new Response(200, $headers, file_get_contents($tempFile));
        unlink($tempFile);
```

Is there a better way that I did not see in the documentation somewhere?

thanks for your help

**Comment by troosan** [comment]
@pavarnos you would want to get the result as a `Psr\Http\Message\StreamInterface`?

**Comment by pavarnos** [comment]
Yes, that would be great (if it is possible?)

**Comment by troosan** [comment]
Indeed, that's what I was thinking, I could add an optional parameter to PhpWord->save() to specify where to stream it to. You can then chose yourself between php://temp, php://memory, ...

**Comment by pavarnos** [comment]
thank you.
I'm currently writing some unit tests for some code that queries a (mock) database to generate a word document file. For an end-to-end (acceptance) test of all the code that goes to make up the document from handling the incoming request to returning a the file, it would also be super handy for the `ReaderInterface` to be able to take a stream as well (it currently can only read files). Then the whole test cycle can be done in memory without touching the disk:
- generate the phpWord document to a StreamInterface with the real code
- in the test, send the stream content to a Reader to rebuild the phpWord structure eg `ReaderInterface->loadFromStream(resource $stream);`
- introspect the phpWord class to look for elements and text runs that contain the expected text

Currently i have to dump the file contents to (yet another) temp file, then read them back in with a Reader.

**Comment by pavarnos** [comment]
for `save()`, what about an interface that separates concerns?
```php
public function saveToFile($filename, $format = 'Word2007');
public function downloadFile($filename, $format = 'Word2007');
public function saveToStream($stream, $format = 'Word2007');
```
where $stream is a string stream name (eg `php://temp`) or an open file handle.

in most cases, `saveToFile` could be as simple as
```php
public function saveToFile($filename, $format = 'Word2007')
{
    file_put_contents($fileName, $this->saveToStream('php://temp', $format));
}
```
for backwards compatibility, your existing `WriterInterface::save($fileName)` could call `saveToFile`.

Maybe. I guess there are some complexities with `\ZipArchive` because it requires a file on disk... These are just random ideas. Feel free to ignore if they don't work.

**Comment by troosan** [comment]
I want to avoid breaking the backward compatibility, so won't modify the existing save function parameters

Have you tried
```php
$phpWord->save('php://temp');
```
You'll have to use the develop branch or change AbstractWriter at line 223
```php
        if (strpos(strtolower($filename), 'php://') === 0) {
```

**Comment by troosan** [comment]
Won't be enough ...
We need to return the resource :-(

**Comment by pwaring** [comment]
I think I'm experiencing the same issue, using Laravel's `streamDownload` function doesn't seem to work with `$writer->save('php://output');`. I'll probably get around it by saving the file locally instead of streaming, but I'm happy to test anything which returns something which works with `streamDownload`.

**Comment by bgmorton** [comment]
Is there any reason not to use output buffering to make it work with Laravel's `streamDownload`?

```

$objWriter = \PhpOffice\PhpWord\IOFactory::createWriter($phpWord, 'Word2007');

ob_start();

$objWriter->save("php://output");

$contents = ob_get_contents();

ob_end_clean();

return response()->streamDownload(function () use ($contents) {
    echo $contents;
}, 'file.docx');

```
